CODING_GUIDELINES.md

```diff
@@ -6,30 +6,82 @@ This document contains the coding guidelines for the ASPIRE Python project.
+Our project follows a reduced [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).
 Please submit any PRs against the `develop` branch.
 
 ![Gitflow Diagram](https://wac-cdn.atlassian.com/dam/jcr:61ccc620-5249-4338-be66-94d563f2843c/05%20(2).svg?cdnVersion=357)
 
+The basic idea is that we have `feature` branches,  `develop` as an integration focused branch, and `master` for releases.
+We do not currently have a need for a dedicated release or staging branch; in our case `develop` also serves this purpose.
+
+Developers are welcome to maintain their own fork, but this is not required.
+
+When `develop` is at a milestone where we want to cut a release we will commit with a PEP440 compliant `Major.Minor` version
+scheme, tag it similarly, and complete a merge to master following the controls below. It is at this point the
+release would be a candadite to upload to PyPI.
+
+###  Git Workflow Controls
+
+To facilitate effective integration, merges into `develop` must occur as a pull request that is approved by two developers.
+One devloper can be anyone, but should probably be someone related to the effort if that is applicable.
+The second developer is a designated codeowner described in the `CODEOWNERS` file.
+
+Additionally, all outstanding requested changes and comments must be resolved before merging.  Pushing new changes should trigger
+a re-review process of changed files to prevent oversights.
+
+Code should be passing integration and unit tests as defined in code on all supported plaforms by the evolving continuous integration systems,
+or be explicitly exempted by a codeowner. This is both for code quality and so maintenence of CI systems is not overlooked.
+
+All of these controls will be managed automatically by the Gitlab server.
+
+(discussion note, maybe want an additional control (Amit?) on master/releases, I don't expect too often..).
+
+### Git Workflow Courtesy
+
+Contributors are responsible for keeping their working repositories up to date with the main Github repo.
+This is particularly important during active development.
+
+If a piece of work is known to change files that will effect other developers' active work, they should be invited to the review or tagged,
+so that the most effective order of integration can be established.  In the case of your own works having integration
+concerns, again, inviting other developers to discuss might be helpful.
+
+In the case a codeowner will not be available, such as vacation/travel, and an emergency patch can not be delayed
+a simple explicit consensus among the remaining developers should suffice.
+(discussion note, This level of urgency seems unlikely, can maybe drop and agree to delay, more likely a codeowner will just be busy,
+which is not as simple).
+
+Do not merge over another persons working branch or pull request unless explicitly permitted to do so by the author.
```

Comment:
This should be a rare situation. But there are events such as GPU hackathons when code reviewing is almost impossible. In this case we should create a separate branch to dedicate it and all updates to this branch should be reviewed and merged after the event.

Reply:
I see. Clearly we still disagree about some of the most basic practices and considerations. That is unfortunate.  I think you are going to need to paint a really clear picture why you specifically need to be able to push over other peoples work.  🍿

Reply:
I'm not sure you guys are talking about the same thing here. @junchaoxia, are you referring to common dev branches like the hackathon branches or to feature branches (like those for ongoing PRs or experimental implementations)? For the former, there's maybe an argument but I don't see why we should allow overwriting other people's work in the latter case.

Would a clearer separation be for branches on forks to be off-limits and branches on the main repo to be carefully coordinated between authors?

Reply:
I think we are.  There's no reason a merge request can't be done for any of them...  It's the same amount of button clicks, but might requiring pinging your coworker. In cases where a deadline is at risk, one might easily argue it is even more critical to do so.

Most projects I've worked, including others I'm working on now, don't require that sort of policy. I think this one does unfortunately.  Fool me once, shrugs.

Reply:
So all branches are considered “owned” by some developer (including `master` and `devel`, which are owned by me) and require permission from that developer for additions and merges (through PRs to those branches). Does that make sense?

Reply:
Yes, if ownership verbiage makes it clearer, sure.

On a collaborative branch, like the hackathon,  I would say in that case it is owned by all involved parties, not just one. I might send PR to Junchao, and he might send his to me, assuming we're the pair collaborating. Sometimes we did that, sometimes we did not. There are also cases where hackathon work was based on a HEAD that was too green and unproven, which really hindered effectiveness and results. Probably this approach would have avoided some of that.  It was better to see these things happen on a blue sky project :).

I think when we have a better synergy we should revisit and relax it a little.  Keep what works, adjust what doesn't.

Reply:
I added 
`Instead, Pull Requests are suggested for all collaborative contributions`.

If that covers the territory we can avoid the ownership verbiage.

Reply:
> There are also cases where hackathon work was based on a HEAD that was too green and unproven, which really hindered effectiveness and results. 

Can you elaborate? There was a lot of untested code merged in too quickly?

> `Instead, Pull Requests are suggested for all collaborative contributions`

Seems a bit too vague.

Reply:
Yes, several untested code changes were merged together. Multiple problems ensued simultaneously. Luckily the main result was ready enough before the event, and this time it wasn't my code.. so I just got to observe and try to help.... We lost the last day or so's worth of work (and at least one major result, which is still on my todo list) because of it. The changes should have been kept orthogonal, testable, and in running condition before integrating.  To take such items now is all rework, probably manual. While post integration was expected, maybe not of that sort/level, at least in cases which could have obviously been decoupled (to me).

I can think about that sentence. I was trying not the have the owner mentality, but maybe that is clearer.

Reply:
Ok so part of the issue might be to ensure that PRs don't get too broad (which should also help reducing the risk of conflicts). Would be good to have something about this in here too.

With regard to ownership terminology, I think it's fine. I mean, if we're going to say that different devs are responsible (“own”) different parts of the code, it might make sense to do the same for branches. That way there's no uncertainty who needs to check off on PRs: if it's to `devel`, it's the code owner, if it's to some other branch, it's whoever owns that branch.

Reply:
In GPU hackathon we need to report a final result including every major steps and the final version mixed with different contributions but for individual branches there are clear differences and they did get mixed up and we did not do PRs to other branches of developers.  Of course there lacks of new unit tests for the GPU version but I think it is acceptable for such a short time.  There are a couple cases in normal PR review process  that Joakim pushed his modifications to my branch but the changes are small and it is easy to handle. For a general rule we are following using the PRs for collaboration.

Reply:
I can add something to that effect. It is hard to gauge, but we can get the idea down at least.

Right, but as far as collaboration branches that would not have avoided the above described problem. Quis custodiet ipsos custodes. With that said, I don't expect this document to fix everything, can refine it later as stuff like that which might come up in practice. Good to try things out...

Reply:
> There are a couple cases in normal PR review process that Joakim pushed his modifications to my branch but the changes are small and it is easy to handle.

Yes and that was probably a bad idea. Better to submit a PR and let you approve it. If it's small and easy to handle, it's just a click away.

> Right, but as far as collaboration branches that would not have avoided the above described problem.

That's true unless you specifically define an owner for that particular branch (it can still be a collaboration, but one person is appointed as responsible for that branch).

Reply:
..Who reviews the owner's changes on a collaborative branch? Just seems like another personal feature branch at that point.

Reply:
Yes. I agree to that we don't need to figure out everything now. Probably a neat version including general rules. That's enough.

Reply:
@junchaoxia , we were unable to do that final report...

We did PRs, just, only sometimes.

Not sure I agree about the unit test remark. For the cufinifft and many changes I added several unit tests which were previously not available and should have been as you suggest... We used them throughout the event.  This is an example of something that rubs me the wrong way. Taking responsibility instead of shedding it is an important component of respect among developers. Stuff happens, shrugs.  If I assume you just didn't know about the additional tests, that is an example of why I think it would be good to review code more. If you we just trying to make a cute remark, it wasn't really helpful to this discussion which I take to be about code gatekeeping.

Yes, in practice it is not uncommon for developers to push commits around etc. In some projects that works effectively.  What I am saying is that did not work well in this project during the short time I have been here to observe it.  We can go into details why that is, but I don't think it is productive.

I'm glad we agree this is enough. I think this is about all I can take, lol.